homeassistant/components/nest/media_source.py

```diff
@@ -0,0 +1,259 @@
+"""Nest Media Source implementation.
+
+The Nest MediaSource implementation provides a directory tree of devices and
+events and associated media (e.g. an image or clip). Camera device events
+publish an event message, received by the subscriber library. Media for an
+event, such as camera image or clip, may be fetched from the cloud during a
+short time window after the event happens.
+
+The actual management of associating events to devices, fetching media for
+events, caching, and the overall lifetime of recent events are managed outside
+of the Nest MediaSource.
+
+Users may also record clips to local storage, unrelated to this MediaSource.
+
+For additional background on Nest Camera events see:
+https://developers.google.com/nest/device-access/api/camera#handle_camera_events
+"""
+
+from __future__ import annotations
+
+from collections import OrderedDict
+from collections.abc import Mapping
+from dataclasses import dataclass
+import logging
+
+from google_nest_sdm.camera_traits import CameraClipPreviewTrait, CameraEventImageTrait
+from google_nest_sdm.device import Device
+from google_nest_sdm.event import ImageEventBase
+
+from homeassistant.components.media_player.const import (
+    MEDIA_CLASS_DIRECTORY,
+    MEDIA_CLASS_IMAGE,
+    MEDIA_CLASS_VIDEO,
+    MEDIA_TYPE_IMAGE,
+    MEDIA_TYPE_VIDEO,
+)
+from homeassistant.components.media_player.errors import BrowseError
+from homeassistant.components.media_source.error import Unresolvable
+from homeassistant.components.media_source.models import (
+    BrowseMediaSource,
+    MediaSource,
+    MediaSourceItem,
+    PlayMedia,
+)
+from homeassistant.components.nest.const import DATA_SUBSCRIBER, DOMAIN
+from homeassistant.components.nest.device_info import NestDeviceInfo
+from homeassistant.components.nest.events import MEDIA_SOURCE_EVENT_TITLE_MAP
+from homeassistant.core import HomeAssistant
+import homeassistant.util.dt as dt_util
+
+_LOGGER = logging.getLogger(__name__)
+
+MEDIA_SOURCE_TITLE = "Nest"
+DEVICE_TITLE_FORMAT = "{device_name}: Recent Events"
+CLIP_TITLE_FORMAT = "{event_name} @ {event_time}"
```

Comment:
We might want to shorten the event names in case the user is on mobile.

> Motion detected @ 12:34

vs

> Motion @ 12:34